### PR TITLE
Add documentation for minimizer tags

### LINKED ISSUE #828: Document the tags
Add documentation that describes the tags used in #782

### PR DESCRIPTION
**Pull request by jess-farmer**
#### Description of Work

Fixes #828 

#### Testing Instructions

1. Proof read docs, make sure they are clear and make sense
2.
3.

Function: Does the change do what it's supposed to?

Tests: Does it pass? Is there adequate coverage for new code?

Style: Is the coding style consistent? Is anything overly confusing?

Documentation: Is there a suitable change to documentation for this change?

### CODE REVIEW

File: docs/source/users/algorithms/line_search.rst [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+.. math::
+    \nabla^2 f(x_k)p = -\nabla f(x_k),
+solve the system
+
+.. math::
+    J_k^T J_k p_k^{GN} = - J_k^T r_k
+
+(where :math:`J_k` is the Jacobian) to obtain the search direction :math:`p_k^{GN}`. The next iterate is then set as :math:`x_{k+1} = x_k + p_k^{GN}`.
+
+Here, the approximation of the Hessian :math:`\nabla^2 f_k \approx J_k^T J_k` has been made, which helps to save on computation time as only second derivatives are not calculated.
```
Comment (AndrewLister-STFC) [inline docs/source/users/algorithms/line_search.rst:107]: Should this have the "only" removed?

File: docs/source/users/algorithms/trust_region.rst [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+.. math::
+    m_k(p) = \frac{1}{2} \|r_k\|^2 + p^T J_k^T r_k + \frac{1}{2} p^T J_k^T J_k p
+
+So, for a spherical trust region, the subproblem to be solved at each iteration is :math:`\min_p \frac{1}{2} \|J_k p + r_k\|^2`, subject to :math:`\|p\| \leq \Delta_k`.
+
+Levenberg-Marquardt uses a combination of gradient descent and Gauss-Newton method. When the solution :math:`p^{GN}` lies inside of the trust region :math:`\Delta`, then :math:`p^{GN}` also solves the sub-problem. Otherwise, the current iteration is far from the optimal value and so the search direction is determined using steepest descent, which performs better than Gauss-Newton when far from the minimum.
+
+**Advantages**:
+    - Robust (more so that Gauss-Newton).
```
Comment (AndrewLister-STFC) [inline docs/source/users/algorithms/trust_region.rst:35]:
```suggestion
    - Robust (more so than Gauss-Newton).
```

File: fitbenchmarking/controllers/bumps_controller.py [hunk 3]
```diff
@@ -11,6 +11,20 @@
 from fitbenchmarking.cost_func.cost_func_factory import create_cost_func
 from fitbenchmarking.utils.exceptions import CostFuncError
 
+ALGORITHM_CHECK = {
```
Comment (AndrewLister-STFC) [inline fitbenchmarking/controllers/bumps_controller.py:14]: As mentioned in the sphinx extension, I think these would be better defined within the class (ideally before the init)

File: docs/source/_ext/algorithmcheckdocs.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+            value = value + software_name + ':' + str(dict[inp_key]) + '\n'
+
+        paragraph_node = nodes.literal_block(text=value)
+        return [paragraph_node]
+
+
+def setup(app):
+    """
+    Setup PrettyPrintModuleValue for use in docs
```
Comment (AndrewLister-STFC) [inline docs/source/_ext/algorithmcheckdocs.py:61]:
```suggestion
    Setup AlgorithmCheckDocs for use in docs
```

**Review by AndrewLister-STFC — DISMISSED** [review]
The docs look great, I spotted a few typos that sliiped in but other than that it looks good on the docs.

I do wonder if we can remove the need for `ALGORITHM_CHECK` as a module level variable though.

File: docs/source/users/algorithms/line_search.rst (hunk 1 above)
Comment (jess-farmer) [reply docs/source/users/algorithms/line_search.rst:107]: Yep, good spot!

File: fitbenchmarking/controllers/dfo_controller.py [hunk 5]
```diff
@@ -32,18 +46,7 @@ def __init__(self, cost_func):
...
-            'deriv_free': ['dfogn', 'dfols'],
-            'general': [],
-            'simplex': [],
-            'trust_region': ['dfols', 'dfogn'],
-            'levenberg-marquardt': [],
-            'gauss_newton': ['dfogn'],
-            'bfgs': [],
-            'conjugate_gradient': [],
-            'steepest_descent': []}
+        self.algorithm_check = self.ALGORITHM_CHECK
```
Comment (AndrewLister-STFC) [inline fitbenchmarking/controllers/dfo_controller.py:49]: This seems unnecessary, is there a reason not to just rename `ALGORITHM_CHECK` to `algorithm_check` and remove this line?

**Review by AndrewLister-STFC — CHANGES_REQUESTED** [review]
I like the new layout without the scroll bars, just 1 last minor thing.

File: fitbenchmarking/controllers/dfo_controller.py (hunk 5 above)
Comment (jess-farmer) [reply fitbenchmarking/controllers/dfo_controller.py:49]: I've tried making this change, but then for all minimizers I get this error message :
`Minimizer cannot be run with Controller with current "algorithm_type" option set.
Details: For the selected software, there are no minimizers with the algorithm type(s) selected in the options file`

So I think that `algorithm_check` in the base controller isn't being overridden?

Comment (AndrewLister-STFC) [reply fitbenchmarking/controllers/dfo_controller.py:49]: Was the default in base_controller still in the __init__? That would override it.

Comment (jess-farmer) [reply fitbenchmarking/controllers/dfo_controller.py]: Ah yes, this was the problem!

File: fitbenchmarking/controllers/base_controller.py [hunk 6]
```diff
@@ -21,27 +21,50 @@ class Controller:
 
     VALID_FLAGS = [0, 1, 2, 3, 4, 5]
 
+    #: Within the controller class, you must
+    #: initialize the a dictionary, ``algorithm_check``,
```
Comment (AndrewLister-STFC) [inline fitbenchmarking/controllers/base_controller.py:25]:
```suggestion
    #: initialize a dictionary, ``algorithm_check``,
```

**Review by AndrewLister-STFC — APPROVED** [review]

**Pull request merged**